### PR TITLE
TP2000-814 Preserve footnote associations on measures edit

### PR DESCRIPTION
**Pull request by dalecannon**
# TP2000-814 Preserve footnote associations on measures edit

## Why
When a measure is updated through the multiple measures edit journey, footnote associations present on the measure aren't being preserved on the new version of the measure because the association is based on the old measure's pk.

## What
- Updates existing footnote associations with the new version of the measure. 
- Adds unit test

## 

Screenshot 2023-03-21 at 14 19 56

### CODE REVIEW

File: measures/tests/test_views.py [hunk 1]
```diff
@@ -1805,6 +1805,83 @@ def test_multiple_measure_edit_only_duties(
...
+        response = valid_user_client.post(url, step_data["data"])
+        assert response.status_code == 302
+
+        assert response.url == reverse(
+            "measure-ui-edit-multiple",
+            kwargs={"step": step_data["next_step"]},
+        )
+
+    workbasket_measures = Measure.objects.filter(
+        trackedmodel_ptr__transaction__workbasket_id=session_workbasket.id,
```
Comment (paulpepper-trade) [inline measures/tests/test_views.py:1873]: I think you can simplify this
```suggestion
        transaction__workbasket=session_workbasket,
```

File: measures/tests/test_views.py [hunk 2]
```diff
@@ -1805,6 +1805,83 @@ def test_multiple_measure_edit_only_duties(
...
+    valid_user_client,
+    session_workbasket,
+):
+    """Tests that footnote associations are preserved in MeasureEditWizard."""
+
+    measure = factories.MeasureFactory.create()
+    footnote_association = factories.FootnoteAssociationMeasureFactory.create(
+        footnoted_measure=measure,
+    )
+    exp_footnote = measure.footnotes.first()
```
Comment (paulpepper-trade) [inline measures/tests/test_views.py:1818]: What's the intent here with using `.first()`? `.approved_up_to_transaction()` perhaps?

File: measures/tests/test_views.py [hunk 3]
```diff
@@ -1805,6 +1805,83 @@ def test_multiple_measure_edit_only_duties(
...
+        trackedmodel_ptr__transaction__workbasket_id=session_workbasket.id,
+    ).order_by("sid")
+
+    complete_response = valid_user_client.get(response.url)
+    # on success, the page redirects to the workbasket page
+    assert complete_response.status_code == 302
+    assert valid_user_client.session["MULTIPLE_MEASURE_SELECTIONS"] == {}
+    for measure in workbasket_measures:
+        assert measure.update_type == UpdateType.UPDATE
+        assert measure.footnotes.first() == exp_footnote
```
Comment (paulpepper-trade) [inline measures/tests/test_views.py:1882]: Same question about `.first()` intent.

**Review by paulpepper-trade — CHANGES_REQUESTED** [review]

File: measures/tests/test_views.py (hunk 2 above)
Comment (dalecannon) [reply measures/tests/test_views.py]: Using `.first()` here returned the first object in the queryset returned by `measures.footnotes`, which allowed for asserting that the footnote association for the old measure and the new version is the same. 

The test has been updated to check for footnote count and to check each footnote in the queryset, not just the first.

**Review by paulpepper-trade — APPROVED** [review]

File: measures/tests/test_views.py (hunk 2 above)
Comment (dalecannon) [reply measures/tests/test_views.py]: Yes, that's right. 

The variable names have now been amended to reflect this more clearly.

**Pull request merged**